Sources/SyncDataProviders/Bookmarks/BookmarksProvider.swift

```diff
@@ -0,0 +1,258 @@
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+//
+
+import Foundation
+import Bookmarks
+import Combine
+import CoreData
+import DDGSync
+import Persistence
+
+public final class BookmarksProvider: DataProviding {
+
+    public init(database: CoreDataDatabase, metadataStore: SyncMetadataStore, reloadBookmarksAfterSync: @escaping () -> Void) {
+        self.database = database
+        self.metadataStore = metadataStore
+        self.metadataStore.registerFeature(named: feature.name)
+        self.reloadBookmarksAfterSync = reloadBookmarksAfterSync
+        syncErrorPublisher = syncErrorSubject.eraseToAnyPublisher()
+    }
+
+    public let syncErrorPublisher: AnyPublisher<Error, Never>
+
+    // MARK: - DataProviding
+
+    public let feature: Feature = .init(name: "bookmarks")
+
+    public var lastSyncTimestamp: String? {
+        get {
+            metadataStore.timestamp(forFeatureNamed: feature.name)
+        }
+        set {
+            metadataStore.updateTimestamp(newValue, forFeatureNamed: feature.name)
+        }
+    }
+
+    public func prepareForFirstSync() async throws {
+        lastSyncTimestamp = nil
+
+        try await withCheckedThrowingContinuation { (continuation: CheckedContinuation<Void, Error>) in
+            var saveError: Error?
+
+            let context = database.makeContext(concurrencyType: .privateQueueConcurrencyType)
+
+            context.performAndWait {
+                let fetchRequest = BookmarkEntity.fetchRequest()
+                let bookmarks = (try? context.fetch(fetchRequest)) ?? []
+                for bookmark in bookmarks {
+                    bookmark.modifiedAt = Date()
+                }
+
+                do {
+                    try context.save()
+                } catch {
+                    saveError = error
+                }
+            }
+
+            if let saveError {
+                continuation.resume(with: .failure(saveError))
+            } else {
+                continuation.resume()
+            }
+        }
+    }
+
+    public func fetchChangedObjects(encryptedUsing crypter: Crypting) async throws -> [Syncable] {
+        return await withCheckedContinuation { continuation in
+
+            let context = database.makeContext(concurrencyType: .privateQueueConcurrencyType)
+
+            var syncableBookmarks: [Syncable] = []
+            context.performAndWait {
+                let bookmarks = BookmarkUtils.fetchModifiedBookmarks(context)
+                syncableBookmarks = bookmarks.compactMap { try? Syncable(bookmark: $0, encryptedWith: crypter) }
+            }
+            continuation.resume(with: .success(syncableBookmarks))
+        }
+    }
+
+    public func handleInitialSyncResponse(received: [Syncable], clientTimestamp: Date, serverTimestamp: String?, crypter: Crypting) async throws {
```

Comment:
We have two places we branch to handle initial sync and deduplication: here and in entity parsing, I wonder if we can make it more unified (e.g. branch only in one place)

Reply:
Disclaimer that we need to stick to 2 public functions, separate for handling initial sync and regular sync. That’s because empty sent array doesn’t mean it’s an initial sync.

We can, however, unify implementations of `handleSyncResponse` and `handleInitialSyncResponse` into a single function handling both scenarios (the only difference between the two is the value of `deduplicateEntities` flag that’s passed to `BookmarksResponseHandler`) which is what I’m going to do.